python/cudf/cudf/core/groupby/groupby.py

```diff
@@ -896,6 +896,114 @@ def corr(self, method="pearson", min_periods=1):
 
         return res
 
+    def cov(self, min_periods=None, ddof=1):
+        """
+        Compute pairwise covariance of columns, excluding NA/null values.
```

Comment:
The docs for pandas' implementation are more extensive. Should we copy some of that information? (This is mostly a question for other reviewers, since I don't know our standards for this.)

https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.cov.html

Reply:
@bdice I would like to know too. 
@shwina Pandas API alignment milestone https://github.com/rapidsai/cudf/milestone/5 might have a guide/plan.

Reply:
Overall we've been tending towards copying in most of the pandas docstrings and linking to pandas webpages when there are non-docstring components to the pandas docs that are more extensive explanations. I don't know if we've codified our preferences though, and it would probably be good to do so.

Reply:
I think we can copy the docstrings with the exception that we should note that we handle nan/nulls differently from pandas. Specifically: everything in the description can be the same. But add a note section below to highlight in cudf the covariance skips any row pairs that contains one ore more nulls. 

Do we know if the result from cudf is always SPD? cc @karthikeyann @bdice

Reply:
Do you mean positive semi-definite? (I usually see it abbreviated PSD, but the abbreviation is uncommon) https://en.wikipedia.org/wiki/Definite_matrix

It's been a while since I read the thread but [this comment](https://github.com/pandas-dev/pandas/issues/16837#issuecomment-313731514) says,

> the functions in both pandas and numpy use the so-called "pairwise deletion" strategy which could lead to non positive semi-definite covariance matrices

If cudf matches pandas, the result is not always positive semi-definite. I'd want to check with some of the input values from that comment to check. (For an easy way to check, remember that [a matrix is positive semi-definite if and only if its eigenvalues are non-negative, >=0](https://en.wikipedia.org/wiki/Definite_matrix#Eigenvalues).)

Reply:
I would definitely want to test this to be sure, but I would expect our result to match pandas and not be PSD in this case. We're explicitly operating on the pairs of columns here, not omitting nulls globally. We could do the latter if we wanted (as I see is discussed elsewhere in this PR), it would just require prefiltering the DataFrame to support those algorithms.

Reply:
I don't know about pair-wise deletion and list-wise deletion, so I looked it up: https://www.ibm.com/support/pages/pairwise-vs-listwise-deletion-what-are-they-and-when-should-i-use-them.

From this, `List-wise deletion` means when you have missing value in an observation for all random variable (basically a row in the dataframe), the entire row should be excluded from the computation. `Pair-wise deletion` means the row should be excluded only when any of the pair of the random variable that's computing has missing values. As an example, consider you have three columns, A, B, C and they look like:
```
A    B    C
1    10   20
<NA> 11   21
2    12   22
```
Since row 2 has a missing value in column A, in list-wise deletion strategy, when computing the covariance of B and C, the row that contains `11` and `21` should be excluded. In cuDF/pandas, we don't see such strategy applied, thus row 2 is taken into consideration. Only when computing A and B, row 2 is excluded, and this is the pairwise deletion strategy that  cuDF and pandas is adopting.

And sure enough, cuDF result matches with pandas in that the result is not PSD.

```python
In [12]: import cudf
    ...: 
    ...: key = cudf.Series([1, 1, 1, 1])
    ...: df = cudf.DataFrame([[1, 2], [None, 4], [5, None], [7, 8]], columns=["v0", "v1"])
    ...: cov_cudf = df.groupby(by=key).cov()

In [13]: cov_cudf
Out[13]: 
           v0         v1
v0   9.333333  18.000000
v1  18.000000   9.333333

In [15]: np.linalg.eigvals(cov_cudf.to_pandas())
Out[15]: array([27.33333333, -8.66666667])
```

Reply:
I think it's worthwhile to add this example into tests to track PSD-ness matches with pandas.

Reply:
> From this, List-wise deletion means when you have missing value in an observation for all random variable (basically a row in the dataframe), the entire row should be excluded from the computation. Pair-wise deletion means the row should be excluded only when any of the pair of the random variable that's computing has missing values.

Right, sorry I should have maybe been more explicit in my explanation. We are constructing our struct columns from pairs of input columns and then running the calculation on that. There is no way that such a calculation could be performing a global filter, it must be performing a pairwise filter. You can see one place where such filtering ultimately occurs [here](https://github.com/rapidsai/cudf/blob/branch-22.04/cpp/src/groupby/sort/aggregate.cpp#L539). It's also fairly straightforward to see why this violates the assumptions that go into proving the definiteness of a covariance matrix because you're no longer computing each of the elements of the matrix using the same data, you're using different data in each case. Since you are no longer computing variances between the same sets of data, the expectation (requirement) of a PSD covariance no longer holds. We _could_ perform list-wise deletion by just doing a dropna before running any of the other code, and that would generate a valid covariance matrix. Whether it's the most useful matrix is really up to the user's needs.

Reply:
Just to clarify, is the agreed decision to leave it as is ( i.e the Pair-wise deletion strategy currently being used, which matches pandas expected beahavior) ... or  to "perform list-wise deletion by just doing a dropna before running any of the other code, and that would generate a valid covariance matrix" ? 
@vyasr @isVoid  @bdice

Reply:
Unless someone feels otherwise I think the action item here is to just document our behavior (as matching pandas) and move on for now, perhaps with a comment to the effect that support for alternative methods of nan handling could be added in the future.

> Do you mean positive semi-definite? (I usually see it abbreviated PSD, but the abbreviation is uncommon) 
As an aside, SPD is indeed also used to refer to these matrices, although it more precisely refers to positive definite (not semi-definite) matrices. In that context SPD means symmetric positive definite. The "symmetric" part is redundant with that abbreviation though since definite matrices are Hermitian by definition.